### PR TITLE
Ordered bijector

### LINKED ISSUE #7: Ordered bijector
Constrain a vector/tensor being ordered along a specific dimension.

### PR DESCRIPTION
**Pull request by junpenglao**
This PR implements the Ordered bijector. The Inverse of Ordered bijector maps a normal distributed vector `y ~ normal(0, 1)` to a sorted random vector that has the same distribution (i.e., `normal(0, 1)`)
close https://github.com/tensorflow/probability/issues/7
cc @srvasude @dustinvtran

### CODE REVIEW

File: tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py]: Copyright 2018

**Comment by martinwicke** [comment]
@srvasude is that the only problem with this PR, or are you reviewing still?

**Comment by martinwicke** [comment]
Np. When you're done, remove the "awaiting review" label.

**Review by srvasude — CHANGES_REQUESTED** [review]
Thanks! Most of my comments are about documentation. This PR looks good, and will be exciting to get in (I'm thinking of writing an example / demo using this for ordinal regression :) ).

**Comment by junpenglao** [comment]
Thanks for reviewing! I only see the comment about copyright, is there anything else?

File: tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+from tensorflow.contrib.distributions.python.ops.bijectors.ordered import Ordered
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops.distributions.bijector_test_util import assert_bijective_and_finite
+from tensorflow.python.platform import test
+
+
+rng = np.random.RandomState(42)
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py]: Prefer to store this as a private member in setUp.

File: tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class OrderedBijectorTest(test.TestCase):
+  """Tests correctness of the ordered transformation."""
+
+  def testBijectorVector(self):
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py]: Can you annotate this and the shape test with   @test_util.run_in_graph_and_eager_modes()?

We are trying to at least make all new tests as eager compatible as possible (this will require changing .eval() to self.evaluate(...).

File: tensorflow/contrib/distributions/python/ops/bijectors/invert.py [hunk 4]
```diff
@@ -18,14 +18,14 @@
...
-from tensorflow.python.ops.distributions import bijector as bijector_lib
+from tensorflow.python.ops.distributions import bijector
 
 __all__ = [
     "Invert",
 ]
 
 
-class Invert(bijector_lib.Bijector):
+class Invert(bijector.Bijector):
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/invert.py]: Thanks for these fixes :).

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    if self.validate_args:
+      is_greater_one = check_ops.assert_greater(
+          output_shape[-1], 1, message="Need last dimension greater than 1.")
+      output_shape = control_flow_ops.with_dependencies(
+          [is_greater_one], output_shape)
+    return (output_shape[-1])[..., array_ops.newaxis]
+
+  def _forward(self, x):
+    x = self._maybe_assert_valid_x(x)
+    y0 = array_ops.expand_dims(x[..., 0], -1)
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Prefer x[...., 0, array_ops.newaxis]. Here and elsewhere.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  bijector.Ordered().inverse([0.06428002, -1.07774478, -0.71530371])
+  # Result: [0.06428002, 0.40464228, 0.8936858]
+  ```
+  """
+
+  def __init__(self,
+               validate_args=False,
+               name="ordered"):
+    self._graph_parents = []
+    self._name = name
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Doesn't need to be set as passed in to super class constructor.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+  bijector.Ordered().inverse([0.06428002, -1.07774478, -0.71530371])
+  # Result: [0.06428002, 0.40464228, 0.8936858]
+  ```
+  """
+
+  def __init__(self,
+               validate_args=False,
+               name="ordered"):
+    self._graph_parents = []
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Not used

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  ```python
+  bijector.Ordered().forward([2, 3, 4])
+  # Result: [2., 0., 0.]
+
+  bijector.Ordered().inverse([0.06428002, -1.07774478, -0.71530371])
+  # Result: [0.06428002, 0.40464228, 0.8936858]
+  ```
+  """
+
+  def __init__(self,
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: nit: all on same line.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+  The inverse of the bijector applied to a normal random vector `y ~ N(0, 1)`
+  gives back a sorted random vector with the same distribution `x ~ N(0, 1)`
+  where `x = sort(y)`
+
+  On the last dimension of the tensor, Ordered bijector performs:
+  `y[0] = x[0]`
+  `y[1:] = math_ops.log(x[1:] - x[:-1])`
+
+  Example Use:
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Prefer

#### Example Use

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+from tensorflow.python.ops.distributions import bijector
+
+
+__all__ = [
+    "Ordered",
+]
+
+
+class Ordered(bijector.Bijector):
+  """Bijector which maps a tensor x_k that has increasing elements in the last
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Is it possible to write the explicit domain and codomain (might make it easier for a user to see when to apply this)? This is also useful since in the future we might be annotation distributions / bijectors with supports and it would be easier to do this from looking at the doc string.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops.distributions import bijector
+
+
+__all__ = [
+    "Ordered",
+]
+
+
+class Ordered(bijector.Bijector):
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Is there a more specific name we can use here? One worry I have is that Order statistics are also a valid bijector (and might be a thing in the future), and so the names can be confused.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    y = array_ops.concat([y0, yk], axis=-1)
+    return y
+
+  def _inverse(self, y):
+    x0 = array_ops.expand_dims(y[..., 0], -1)
+    xk = math_ops.exp(y[..., 1:])
+    x = array_ops.concat([x0, xk], axis=-1)
+    return math_ops.cumsum(x, axis=-1)
+
+  def _inverse_log_det_jacobian(self, y):
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Can you add as  comments here and below a proof / summary of a proof of these results?

It would be helpful for debugging purposes / if someone wants to go back to see the implementation (in this case, the jacobian matrix is lower triangular, which is why you only get a sum of the log of the diagonal entries).

**Review by srvasude — CHANGES_REQUESTED** [review]
3rd time is the charm.

File: tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py (hunk 2 above)
Comment (junpenglao) [reply tensorflow/contrib/distributions/python/kernel_tests/bijectors/ordered_test.py]: Do you have an example? I copied this from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/kernel_tests/bijectors/softmax_centered_test.py#L31

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 9 above)
Comment (junpenglao) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: so without the `:`?

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 11 above)
Comment (junpenglao) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: I would suggest call this the `Ordered` bijector as it is used similarly in Stan and PyMC3. The other variation of ordered statistics maybe better to call it a `Sorted` bijector?

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 10 above)
Comment (junpenglao) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Both the domain and codomain is R_k, however, in practice, the inverse mapping usually apply on y~Normal(0, 1), as y_k too large will cause overflow in exp(y_k).

Should I add the above to the description?

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 9 above)
Comment (srvasude) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Hmmm I guess the comments use markdown. I was suggesting using 4 pound signs before Example Use, so that the generated doc would reflect that.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    return math_ops.reduce_sum(y[..., 1:], axis=-1)
+
+  def _forward_log_det_jacobian(self, x):
```
Comment (srvasude) [inline tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Also add _maybe_assert_valid_x(x) here.

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 11 above)
Comment (srvasude) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Yeah I guess that is my worry, which is Sorted and Ordered sound very similar (and I wouldn't trust myself to know which one to use :p). If we can make those sound different enough, make it easy for someone new to this stuff understand this bijector, and also make it easy for those using Stan / PyMC3 to know what this is, I'd be happy (I admit  this is an unfair question, in that I haven't provided any suggestions, but am currently thinking about a name, and will post back).

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 9 above)
Comment (junpenglao) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: Sorry I do not quite get it...  you mean a `$`? An example would be greatly appreciated!

Comment (srvasude) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: \#\#\#\#\# Example Uses

File: tensorflow/contrib/distributions/python/ops/bijectors/ordered.py (hunk 11 above)
Comment (junpenglao) [reply tensorflow/contrib/distributions/python/ops/bijectors/ordered.py]: It is not exactly the same thing as Ordered logistic (as in ordered logistic distribution ordered logistic regression) though - The ordered vector returned from this bijector is used to generate ordered logistic regression and ordered probit regression, so I think naming it Ordered bijector shouldnt be too confusing.

**Comment by junpenglao** [comment]
@srvasude I am not sure if my changes are fully satisfactory, could you provide some further comments?

**Comment by srvasude** [comment]
Apologies, thought the comments will auto-close. Taking a look now.

**Review by srvasude — DISMISSED** [review]
Thanks for the changes (and sorry for the slow review)!

Code changes look good. The name Ordered is fine as is.

**Comment by junpenglao** [comment]
Thanks for reviewing! @martinwicke

**Review by martinwicke — APPROVED** [review]
